### ISSUE TITLE
Conflict of interest setting discount for SegWit

### ISSUE THREAD

**Issue by btcdiscounted** [btc1/bitcoin#92]
Given that SegWit seems to be going into Bitcoin, I would like to know how the conflict of interest in setting the discount for witness data is going to be handled. It seems that the only real way to solve it would be to have government regulators decide the discount, which would be against Bitcoin's ethos.

The discount alters the type of transactions miners are economically incentivized to include in the next block. By changing the discount one can effectively exercise economic policy over the Bitcoin network. If that is not enough, it seems the developers are the ones in charge at the moment of deciding the amount of the discount, but these same developers work for companies that benefit from having a certain discount. For example, @jgarzik , leader of the implementation most miners run, has a side chain company that benefits if witness heavy transactions are cheaper. This is not a personal attack on @jgarzik , but even if one believes he is perfectly honest in his intentions, the problem remains that the conflict of interests should not exist. The same applies to other developers, for example from Blockstream, which also benefit if witness heavy transactions are cheaper.

And even if someone believes that Bitcoin can get away with having such conflict of interest, apart from having dubious morals, there is the problem that having the discount opens a door for regulators to start gaining power in Bitcoin. It's very hard to justify to the population (mainly because they have a point) that some individuals working for some private companies are going to decide parameters that have a strong influence on transactions prices, specially when those private companies have an interest in a certain direction. It's very easy to make the case then that Bitcoin needs regulatory oversight. Once regulators are making economic policy by forcing miners to run a tweaked parameter, it's very easy to sell the need to increase the control.

It would also be hard to sell to new users why they should change from a currency where economic policy is made by policy makers, to another currency where economic policy is made by private companies.

The origin of the problem is the discount is a point of centralization and goes against Bitcoin's ethos. The discount is an economic parameter, as it affects prices. It is not purely an engineering parameter, like network latencies for example, but is mainly an economic parameter. By allowing developers to centrally set economic parameters you are creating a conflict of interest and opening the door for regulatory intervention in Bitcoin.

I understand some people will say "just run another client with a different discount", but that is a very weak excuse, as not accepting blocks with the consensus discount would mean forking away. Even if the answer is to say tthe majority has decided the to run with the discount, that answer is equivalent to saying: "vote a different president to get a different Federal Reserve chairman". This situation is something Bitcoin wanted to avoid, not recreate.

The obvious solution would be to make the weight neutral again, and work to remove other centralized economic parameters in bitcoin and replace them with an algorithm that does not require developers or regulators intervention.

**Comment by ptschip** [comment]
On 19/07/2017 8:21 AM, btcdiscounted wrote:
> The obvious solution would be to make the weight neutral again

This makes perfect sense.    Giving a discount for off chain txns is
nothing more than a marketplace distortion.  It's analogous to some sort
of goverenment subsidy or tax break for certain companies;  in the end
it amounts to someone or group deciding which businesses win and which
lose rather than letting the free market prevail...

**Comment by christophebiocca** [comment]
There's been some discussion in the mailing list: https://lists.linuxfoundation.org/pipermail/bitcoin-segwit2x/2017-June/000063.html

In practice the weight discount only has a measurable impact if blocks are at their limit. If the limit is above demand, then miners can price tx however they want, and orphaning risk (which has been argued to be the driver of fee-level) is proportional to total tx transmission size, with no discounting whatsoever.

**Comment by betawaffle** [comment]
Wow, let me try to dispute as much of this as I can.

>I would like to know how the conflict of interest in setting the discount for witness data is going to be handled. It seems that the only real way to solve it would be to have government regulators decide the discount, which would be against Bitcoin's ethos.

The only people deciding anything are those running full nodes. All the core developers can do is make recommendations (so far based on actual research) and implement those recommendations. They can't and don't force anything on the network, the network decides.

>By changing the discount one can effectively exercise economic policy over the Bitcoin network.

That's no more of an effective argument than if I said changing the block reward one can effectively exercise economic policy over the Bitcoin network. The trick is you'd need to get everyone to adopt such a change.

>... the problem remains that the conflict of interests should not exist.

If we take the step of assuming such a conflict does exist, it still doesn't enable any of the developers to foist upon the network a change the network doesn't agree with.

>Once regulators are making economic policy by forcing miners to run a tweaked parameter ...

Forcing miners to run a tweaked parameter (if it was even possible) wouldn't cause any of the things you're claiming. Miners can't force anything on the network more than anyone else, there always has to be agreement.

**Comment by Pheromon** [comment]
> The only people deciding anything are those running full nodes. All the core developers can do is make recommendations (so far based on actual research) and implement those recommendations. They can't and don't force anything on the network, the network decides.

So you agree that there should be no discount: if a miner want to tweak its settings, it will do it, no need for central planners to let some bytes count more (or less) than others.

1 byte is 1 byte.

**Comment by mjamin** [comment]
The discount is a way of increasing the blocksize without a hard fork.

In a hard fork scenario, if you don't upgrade you are shit out of luck.

In the SegWit soft fork scenario, if you don't upgrade you _maybe_ have to pay a slightly higher fee than upgraded nodes do in some circumstances.

Why is the latter worse than the former?

**Comment by betawaffle** [comment]
>So you agree that there should be no discount: if a miner want to tweak its settings, it will do it, no need for central planners to let some bytes count more (or less) than others.

That's not possible without changing the consensus rules. They would have to convince the entire community to do so.

**Comment by Pheromon** [comment]
> That's not possible without changing the consensus rules. They would have to convince the entire community to do so.

fee discount for witness is not a consensus rule: every miner can decide the fees that he wants to include the tx on his block.

**Comment by Pheromon** [comment]
> The discount is a way of increasing the blocksize without a hard fork.

I know, but segwit2x is an hard fork.

And, btw, if you want to make it successfull the block increase must be done together with segwit activation, otherwise all the miners that now are false signaling bit 4 only to activate segwit will not follow the hard fork.

**Comment by betawaffle** [comment]
>fee discount for witness is not a consensus rule: every miner can decide the fees that he wants to include the tx on his block.

Yes, miners can decide that, but they can't *change* the weight of transactions. If they choose to not include a transaction, they are giving up fees. They are allowed to do this, but it wouldn't be very rational.

**Comment by kek-coin** [comment]
This has been discussed before and iirc the conclusion was "out of scope". @jgarzik correct me if I'm wrong.

**Comment by jgarzik** [comment]
I'm very sympathetic as this remains controversial; it's worth addressing in a blog post at a minimum, as it wraps together total block size, block weight, and related details into one package.

Feel free to continue the discussion here - closing the issue as it's not something we will likely address.

**Comment by mpatc** [comment]
I for one think the block weight should be neutral. I've tried to find a
reason why the 75% number is there, and it seems to be arbitrary. Why not
80% discount? Or 70%. There is no mathematical reason I see 75% was chosen.


For this reason, I think the discount (if any) should at least be dynamic.
Or simply neutral.

In the long term, pruning solutions seems a much better strategy than the
convoluted, arbitrary witness discount.

On Jul 19, 2017 10:15 PM, "Jeff Garzik" <notifications@github.com> wrote:

I'm very sympathetic as this remains controversial; it's worth addressing
in a blog post at a minimum, as it wraps together total block size, block
weight, and related details into one package.

Feel free to continue the discussion here - closing the issue as it's not
something we will likely address.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/btc1/bitcoin/issues/92#issuecomment-316576501>, or mute
the thread

**Comment by mjamin** [comment]
@Pheromon:

> And, btw, if you want to make it successfull the block increase must be done together with segwit activation, otherwise all the miners that now are false signaling bit 4 only to activate segwit will not follow the hard fork.

In other words you want to block SegWit deployment in order to use it as a bargaining tool to get what you want, despite the fact the the network as a whole obviously doesn't want it in the form you propose.

Why do you think that's OK?

**Comment by JaredR26** [comment]
> despite the fact the the network as a whole obviously doesn't want it in the form you propose.

Pot, meet kettle.

**Comment by Pheromon** [comment]
@mjamin it seems like you're good at getting things backward.
There is a strong consensus to have both segwit and a block size increase.
I argue that the plan to have them in separate activation time gives the small minority a tool to enable the former and block the latter.
History will tell.

**Comment by drunkenclam** [comment]
> I for one think the block weight should be neutral. I've tried to find a
> reason why the 75% number is there, and it seems to be arbitrary. Why not
> 80% discount? Or 70%. There is no mathematical reason I see 75% was chosen.

It's not arbitrary. If you had tried googling for a few seconds, you would have found something like [this](https://segwit.org/why-a-discount-factor-of-4-why-not-2-or-8-bbcebe91721e).

**Comment by mjamin** [comment]
@JaredR26

> Pot, meet kettle.

What makes you think I'm being hypocritical? Essentially everybody is in favour of SegWit. What SegWit2x tries to do is bundle something everybody wants (or can accept) with something only a presumably tiny subsection of the ecosystem wants. If you dive into the question of why they want it the way they want it, it's usually answered with nothing more than "because".

If you want to hard fork, then simply create a client that hard forks and see how much support you get. Don't use SegWit as a carrot on a stick.

@Pheromon 

> it seems like you're good at getting things backward.

It might look that way from your perspective.

**Comment by Pheromon** [comment]
@mjamin 
> Essentially everybody is in favour of SegWit

If this would be the case, we would have segwit since a long time, don't you agree?

> It might look that way from your perspective.

See above: my perspective is grounded on reality. And reality show that segwit is a contentious feature since it changes radically the Bitcoin security: the chain of signatures that defines a bitcoin is no more.

So, there are people that want to continue to use Bitcoin as designed by Satoshi (with its transactions and big blocks), and people that want to use a different security model to move transactions offchain (thus taking away incentives from miners and giving them to LN node operators).

Segwit2X is trying to have both in Bitcoin without forking it in two different coins: the original, and the altcoin with segwit and artificial limitations on block size.

**Comment by mjamin** [comment]
> And reality show that segwit is a contentious feature since it changes radically the Bitcoin security: the chain of signatures that defines a bitcoin is no more.

It's not contentious by any meaningful measure. a) https://en.bitcoin.it/wiki/Segwit_support and b) in the past 24 hours, 80% of the hashrate signalled for it via BIP91.

> So, there are people that want to continue to use Bitcoin as designed by Satoshi

Satoshi himself had the core idea behind the Lightning Network: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2013-April/002417.html

So the Lightning Network is undeniably part of "Satoshi's vision" (something that big blockers like to invoke as a killer argument).

> Segwit2X is trying to have both in Bitcoin without forking it in two different coins: the original, and the altcoin with segwit and artificial limitations on block size.

SegWit2x is an attempt to save face and a setup to play the victim in 3 months to keep the anti core narrative going.

**Comment by btcdiscounted** [comment]
Answering to @betawaffle 

> The only people deciding anything are those running full nodes. All the core developers can do is make recommendations (so far based on actual research) and implement those recommendations. They can't and don't force anything on the network, the network decides.

This is a coup out. By parts:

1. The parameter has been decided by Core developers and accepted by miners and bitcoin big business. Anyone else has had no saying and the only option if they disagreed is to leave Bitcoin. That is the reality, saying anything else is lying.

Going forward, the parameter will be changed based on the interests of a few people, developers, miners and maybe big bitcoin business.That is bad design.

Also, non mining full nodes is a sybill-able count and has no value. I don't understand the fetishism for non-mining full nodes that some people have when its a meaningless statistic.

And if you think this coup out is going to work once regulators come in, you are deluded. Once you have a few private business deciding economic parameters in Bitcoin, regulators are going to claim public interests and start to regulate it. Nobody will protest because Bitcoin will still work and the only solution you'll have if you don't want to follow the regulators is fork off with your meaningless non-mining full node. But this excuse of "it was the community who accepted it" is going to be laughed at (and rightly so), because it is obvious that a few private companies with direct interests are setting the parameter.

2. There is no valid science behind the value of 4 for the discount. The only thing that has been pointed out is this link https://segwit.org/why-a-discount-factor-of-4-why-not-2-or-8-bbcebe91721e . That is a piece of propaganda. I understand the need for propaganda, but, to the person who posted that link, please don't bring propaganda to a proper discussion if you don't want to be taken for a fool.

The reality is the discount is supposed to translate to the transaction price the cost of the utxo vs the cost of the witness data. That is its economic role in the model. Taking only this into consideration, there is no justifiable reason as to why it should be 4 and not 3.5 or 4.5 or any other number. Its completely made up. Too big because some statistical model of past transactions said it, is not a valid reason. Which takes me to the obvious problem.

The obvious problem is the believe that has been harbored by more and more Bitcoin developers that economic parameters can be modeled. They can not, central planning does not work. And if someone believe they can, they should go try reform the Federal Reserve, instead of being involved in Bitcoin. Tell congress how your amazing engineering model can find the correct interest rates for the nation. But trying to introduce economic parameter under the assumption that you can model the economic needs of the future and come up with a right number is anti-Bitcoin.

Introducing the discount was lazy developing. What should have happened is that an algorithm based on crypto and economic incentives should have governed the trade off between the utxo cost and the witness cost. But lazy developers instead introduced a centralization point by introducing a new economic parameter (same developers that claim to worry so much about centralization btw). If you think there is no need for this algorithm and central economic parameters are ok in bitcoin, why not get rid of the fee market, set up a minimum fee updated regularly by the models of developers and accept transactions as first comes first goes? The reason is we know the economic needs of the future are unknown and changing, so a market with correctly aligned incentives produces the best results while at the same time avoiding politics.

It is not random that the backdoor meetings between key players of the Bitcoin ecosystem has started once economic parameters came into play. Governing these economic parameters gives power in the Bitcoin network. This politicking is only going to intensify as long as we have economic parameters decided in a central way in Bitcoin. Eventually, the regulators will take over them. It is sad to see because Bitcoin was supposed to be an open system where everybody was the same because the network was governed by a set of algorithms with the right economic incentives. Not anymore apparently.

>>By changing the discount one can effectively exercise economic policy over the Bitcoin network.

>That's no more of an effective argument than if I said changing the block reward one can effectively exercise economic policy over the Bitcoin network. The trick is you'd need to get everyone to adopt such a change.

The block reward is a constant. Any constant would be acceptable. Instead of starting at 50, it could have started at 500, 1, a billion or 0.000001. There would have been no difference to the bitcoin economic model. The discount regulates the cost between utxo and witness data. A different value has direct economic consequences.

I fail to see how they are remotely similar. This is very basic.

 >Forcing miners to run a tweaked parameter (if it was even possible) wouldn't cause any of the things you're claiming. Miners can't force anything on the network more than anyone else, there always has to be agreement.

But they do. If regulators come in and claim for the public good that they are going to decide these economic parameters that are being introduced in Bitcoin, people are not going to rebel because the system will keep working very similarly. At least most users wont rebel as long as things keep working more or less similarly. You can fork off with your meaningless non-mining full node if you want out of principle and be happy, but the majority of the network will accept the regulators directives.

**Comment by christophebiocca** [comment]
For what it's worth, *increasing* the witness weight from 0.25 to 0.5 or 1 (or 2 :P) merely requires a soft-fork.

So it doesn't have to be addressed at hard-fork time unless there's a desire to set an even higher base-block-size (which is probably out of scope).

**Comment by betawaffle** [comment]
>The parameter has been decided by Core developers and accepted by miners and bitcoin big business. Anyone else has had no saying and the only option if they disagreed is to leave Bitcoin. That is the reality, saying anything else is lying.

If you believe that, then Bitcoin has already failed at its goals. I don't, but if it's true, Bitcoin was a failure.

>Also, non mining full nodes is a sybill-able count and has no value. I don't understand the fetishism for non-mining full nodes that some people have when it's a meaningless statistic.

The *number* of full nodes don't matter. What matters is who's using them. A full node is the only way you can guarantee you're actually receiving Bitcoin that conforms to rules you agree with. If you don't run a full node, you're doing little more than trusting the sender. Without a full node, Bitcoin isn't *trustless*.

>Once you have a few private business deciding economic parameters in Bitcoin ...

That's why we mustn't capitulate to them. What are you even suggesting?

>There is no valid science behind the value of 4 for the discount.

There are many reasons it was chosen. Let me see if I can give an overview of some of them:
- A transaction consists of many costs, miners only need to receive and validate it once, but full nodes have to store and index it *at least* until it is later spent. The pre-segwit costs favor short-term miner profits over long-term node operation. A decision which is all the more obviously foolish since compact blocks removed most of the transaction costs from the critical path.
- The witness "discount" doesn't benefit any particular kind of transaction except transactions whose outputs get spent.
- What the weight effectively does is shifts some of the cost from _spending a coin_ in the utxo set, to _creating a coin_.  If you ignore the capacity increase (which lowers them), the total cost for a lifecycle of a coin is the same before and after. The cost is moved to creation time.
- Moving it up disadvantages parties who create outputs but never spend them, which I think is objectively the right move, because keeping those outputs around has a perpetual cost for every node.

>But trying to introduce economic parameter under the assumption that you can model the economic needs of the future and come up with a right number is anti-Bitcoin.

Bitcoin has always had decisions baked in that affect the economics of it. What makes you think the original choices are any better than future ones? Satoshi himself made many such changes as more information became available. Everyone involved in this research space has more information now on the behavior of the system.

>What should have happened is that an algorithm based on crypto and economic incentives should have governed the trade off between the utxo cost and the witness cost.

That is *exactly* what they did. What makes you think otherwise?

>Why not get rid of the fee market?

Nobody can just *decide* there is to be no market. Markets create themselves.

>It is not random that the backdoor meetings between key players of the Bitcoin ecosystem has started once economic parameters came into play. Governing these economic parameters gives power in the Bitcoin network.

Again, what makes you think everyone will accept economic parameters they don't agree with? It's as if you really don't understand how Bitcoin protects itself from such things. If everyone gives up on trustlessness, of course you'll be at the whim of whoever you put your trust in (ie. companies or the government).

You seem to want one thing (the same things I want), then argue against the way we can get there.

>It is sad to see because Bitcoin was supposed to be an open system where everybody was the same because the network was governed by a set of algorithms with the right economic incentives. Not anymore apparently.

It only ever works that way if we run truly validating nodes and don't let people walk all over us.

>... governed by a set of algorithms with the right economic incentives.

Those algorithms and incentives **are what a Bitcoin node does**. You can't have them without one.

>The block reward is a constant. Any constant would be acceptable. Instead of starting at 50, it could have started at 500, 1, a billion or 0.000001.

If you tried to change the block reward (and got away with it) that would have a huge economic impact.

>The discount regulates the cost between utxo and witness data.

The *point* of the change is to *correct* the incentives to make running nodes less costly. As I said before, everyone who can't run a full node is missing out on the benefits of Bitcoin, and is at the mercy of everyone else's decisions.

> If regulators come in and claim for the public good that they are going to decide these economic parameters that are being introduced in Bitcoin, people are not going to rebel because the system will keep working very similarly. At least most users won't rebel as long as things keep working more or less similarly. You can fork off with your meaningless non-mining full node if you want out of principle and be happy, but the majority of the network will accept the regulator's directives.

What do you think is preventing regulators from doing so right now (pre-segwit)? What are you suggesting we do to prevent it without running full nodes?

**Comment by Pheromon** [comment]
@mjamin 

> It's not contentious by any meaningful measure.

So, explain me why it has not even reached 40% of miner support before the segwit2x proposal :-)

> in the past 24 hours, 80% of the hashrate signalled for it via BIP91.

Yes, and you should know that BIP91 has been made expressly because 80% is the hashrate required by the segwit2x initiative. Otherwise segwit would be enabled on bit1 client but not on core!

> Satoshi himself had the core idea behind the Lightning Network: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2013-April/002417.html

The LN network is very cool and I hope it can come to reality.
I, like many other so called "big blockers" am not against the LN (quite the contrary), I'm against an artifcial limit on the block size, that is kept in place only to _force_ users to use the LN because unable to use onchain transactions.

LN has to compete with onchain transactions with its own merits, not because some central authority decided that only a very small percentage of all users can transact onchain, even if technically possible.

> So the Lightning Network is undeniably part of "Satoshi's vision" (something that big blockers like to invoke as a killer argument).

I hope this clears the things a bit for you.

**Comment by betawaffle** [comment]
>... that is kept in place only to force users to use the LN because unable to use on-chain transactions.

Bigger blocks force users out of the ability to run full nodes, which is the only thing protecting them from outside control.

**Comment by Pheromon** [comment]
> Bigger blocks force users out of the ability to run full nodes, which is the only thing protecting them from outside control.

That's completely false, and your sentence has no meaning without specifying how big the blocks.
Are you suggesting that with 1MB blocks I can run a node, and with 1.1MB blocks I'm suddendly unable to?

Technology evolves and the cost of running a node which processes bigger and bigger blocks in time is decreasing each year.

Just think that right now I can run a full node on a raspberry that costs like the fees of a few bitcoin transactions...

Full nodes are useful for most of business but not for normal (non power) users, and every kind of business (and power users) can have a pc costing a few hundreds dollars to run it.

**Comment by christophebiocca** [comment]
> Bigger blocks force users out of the ability to run full nodes, which is the only thing protecting them from outside control.

Outside control is trivial through soft-forks, including soft-forks that completely rewrite the near-entirety of bitcoin's rules. I can give you examples if you're interested.

The only rules not subject to change through soft-forks are rules that affect data in the headers (usually, there are exceptions even then), but such rules are also checked by SPV clients.

**Comment by betawaffle** [comment]
>Are you suggesting that with 1MB blocks I can run a node, and with 1.1MB blocks I'm suddenly unable to?

We have already excluded people with 1MB blocks. It's a sliding scale, the larger the blocks, the more people won't be able to afford running a node. We are already going to lose people with the increase that segwit brings.

**Comment by betawaffle** [comment]
>Outside control is trivial through soft-forks, including soft-forks that completely rewrite the entirety of bitcoin's rules. I can give you examples if you're interested.

Please do.

**Comment by mpatc** [comment]
Dear Andrew Hodges:

I am confused, I hope you can clarify. You said:

> There are many reasons it was chosen. Let me see if I can give an
overview of some [...]


None of the reasons you give for the discount equal 4.  You give some great
reasons for a discount *in general*, but none that I can see for using a
value of 4.

Why not 3 or 5.6 or any other number?

Thanks!
Apologies for formatting, I'm on mobile.

**Comment by betawaffle** [comment]
>None of the reasons you give for the discount equal 4.  You give some great
reasons for a discount *in general*, but none that I can see for using a
value of 4.

Ahh, sorry.

It was the best they could do without decreasing the apparent "legacy" block size without also making segwit blocks too big. They determined that 4MB was the most they could responsibly implement, and the discount was as high as they could make it fitting those parameters. They wanted a higher discount, but it would have meant a larger *real* block size.

**Comment by christophebiocca** [comment]
Sure, here's a few

## Empty chain

A block is invalid if:

- It includes any non-coinbase tx.
- The coinbase tx has a non-zero amount.

Net effect: Bitcoins can never be earned nor sent.

## Inner merkle-tree (AKA the no-op forced upgrade).

In addition to the above rules, the coinbase tx must include a commitment to a new merkle tree that obeys the usual bitcoin rules. People who had regular bitcoin have the ability to use it in this inner block.

Net effect: People must upgrade their software to earn or spend Bitcoin.

### Communist revolution

Same as no-op except the starting inner UTXO is altered such that each address now has the same amount of coins.

### Permaflation

Same as no-op except the reward halving rule is removed.

## Demurrage

This one doesn't require an inner-ledger trick:

All tx must include a miner fee of 1% of their value for each year since the output was created.

## Inner header (AKA hard fork as a soft fork)

Instead of a merkle tree commitment, require the hash of a new type of header.

### Semi-federated sidechain

The inner-header must include a valid signature of a set of pre-ordained keys.

### Additional proof-of-work

The inner-header must have valid proof of work using SHA3, which has a target time of 5.5 blocks/hour.

These two are actually less interesting because they cannot obviate the need for proof of work fully, due to the requirement of maintaining the longest chain (and orphaning any minority chain).

Then there's much more hacky techniques that change even [block interval](https://www.reddit.com/r/btc/comments/428tjl/softforking_the_block_time_to_2_min_my_primarily/). That one's particularly interesting because it requires no cooperation from anyone but the mining majority while changing a rule that literally everyone is supposed to check.

**Comment by betawaffle** [comment]
@christophebiocca At least some of those aren't very realistic. Also, even soft forks can't happen unless people actually run them.

**Comment by Pheromon** [comment]
@christophebiocca great list

Only, I fail to see how "communist revolution" and "permaflation" are soft forks: blocks generated with those rules would be rejected by older (non-upgraded) clients.

**Comment by christophebiocca** [comment]
> At least some of those aren't very realistic.

I went for comedy. These are all 100% doable, but no sane miner would do them. That doesn't mean that the reason they can't happen is *people running full nodes*.

> Also, even soft forks can't happen unless people actually run them.

If by people you mean *miners* yes. If by people you mean *non-mining full-nodes* then no.

> I fail to see how "communist revolution" and "permaflation" are soft forks: blocks generated with those rules would be rejected by older (non-upgraded) clients.

Because the older clients (and any minority miners) don't see the inner ledger at all, and therefore can't enforce any rules in it.

This is similar to how segwit works around the blocksize limit by creating a new commitment to new data that lives outside the data structure validated by older clients. Old clients don't stop it because it's invisible to them. New clients, of course, can be arbitrarily different.

From the perspective of old clients and miners, none of their tx ever confirm, and all of their blocks get orphaned.

**Comment by betawaffle** [comment]
>I went for comedy.

Glad to see you have a sense of humor, but if you hadn't said that people might have believed these as something miners could do without risk.

**Comment by christophebiocca** [comment]
But can we agree that "people running full nodes" is not what prevents all of these forms of miner-imposed new rules from happening?

**Comment by betawaffle** [comment]
>But can we agree that "people running full nodes" is not what prevents all of these forms of miner-imposed new rules from happening?

Yes, but only to a point. If 90% of nodes don't enforce new miner rules, the miners who defect can earn more money. And obviously if miners did something like "empty chain" there would be an enormous incentive to defect.

**Comment by betawaffle** [comment]
Other miner-created soft forks that don't cause consensus issues are fine, btw. If they find a way to make additional money, they are welcome to do it.

**Comment by christophebiocca** [comment]
> If 90% of nodes don't enforce new miner rules, the miners who defect can earn more money.

Miners who defect would be orphaned. They would earn nothing.

**Comment by betawaffle** [comment]
>Miners who defect would be orphaned. They would earn nothing.

It really depends on the situation. If 50% of miners are truly colluding in that way, we've already got a problem. Of course it would cause turmoil, but all miners vs all users never works out well for anyone.

**Comment by christophebiocca** [comment]
> If 50% of miners are truly colluding in that way, we've already got a problem.

Yup, but if 50% of miners are **not** colluding in that way, then SPV is safe from outside control. Which is why your original assertion:

> run full nodes, which is the only thing protecting them from outside control.

Is incorrect.

What a full node has that an SPV node doesn't have, is protection against tx inside fraudulent blocks. This *is* nice but if someone's making destined-to-be-orphaned blocks to defraud you that's a big chunk of change they're spending.

**Comment by betawaffle** [comment]
Correct. SPV provides more security than nothing at all, but it doesn't protect you from unwanted rule changes.

**Comment by christophebiocca** [comment]
>  but it doesn't protect you from unwanted rule changes.

Neither does a full node? We just went over this a minute ago.

**Comment by mjamin** [comment]
@Pheromon:

> So, explain me why it has not even reached 40% of miner support before the segwit2x proposal :-)

You do realize that SegWit2x is the same SegWit as BIP141, just activated differently (lower threshold over a shorter lock-in period) and with an implementation of a hard fork _proposal_ tacked on?

Everybody who supports SegWit2x, also supports -- by definition -- SegWit in general. Everybody who is willing to run SegWit2x but not SegWit as implemented by core is just trying to get a hard fork through by using SegWit as a bargaining tool.

Funny thing is that thanks to BIP91 it doesn't really matter. SegWit will activate and the suggested hard fork will simply not happen as people will continue to use core and the miners and businesses trying to push it will chicken out.

This will be my last reply here, so don't bother asking me further questions.

**Comment by betawaffle** [comment]
In aggregate it does, if the rules you want coincide with most other bitcoin users.

**Comment by NiKiZe** [comment]
@mjamin There is a big difference between "agree with" and "concede to"
~50% of those that runs Segwit2x only agrees on Segwit at all if there also is a blocksize increases (which is implemented and not only proposed)

There will be a fork since miners are already running the btc1 codebase which has the enforced 1MB+ blocksize enforcment. And even before the btc1 block enforcement kicks in there will be another one.

So as shown above - your claims are wrong.

**Comment by Pheromon** [comment]
> Everybody who supports SegWit2x, also supports -- by definition -- SegWit in general.

Let me explain you the word "compromise".
If I do not like segwit but if I want bigger blocks and I realize that this will not happen alone, I can concede to have it, *if* this will also raise the block size.

> Everybody who is willing to run SegWit2x but not SegWit as implemented by core is just trying to get a hard fork through by using SegWit as a bargaining tool

That's exactly my point.

**Comment by betawaffle** [comment]
Except that this bargaining tool is non-binding :rofl:

**Comment by Pheromon** [comment]
> Except that this bargaining tool is non-binding 

I totally agree, and that's exactly why, in a previous message, I stated that the only way to make sure s2x successful is to bind the hardfork with segwit activation.

**Comment by JaredR26** [comment]
@mjamin 

> What makes you think I'm being hypocritical? Essentially everybody is in favour of SegWit.

This part, right here.  The raw facts of the issue, no matter what way you look at it, are that segwit never had the consensus for activation by itself.  When core "declared consensus" on December 19th-23rd 2015, a lot of core devs either already had or soon thereafter quit over the issue.  There were far more posts, commits, as well as more unique posters before the "consensus" than after.  The situation was so bad that one month later 5 core devs met with miners, 72% of whom had signaled for classic at one point in time, to try to derail classic.  That was the HK agreement 1.0.  It specifically called for a +2mb _base block hardfork_, but afterwards I literally only found one dev referencing that on the email list, one time.  No attempt to deliver on that part of the agreement was made, and every core dev who signed that agreement opposes segwit2x(Less Cory Fields who isn't active anymore).  And no, segwit doesn't count as "the agreed block size increase," as the agreement was crystal clear - 2mb base block hardfork.

So what about since then, surely segwit had consensus, after all "essentially everybody is in favor of it" right?  The only place you find that opinion is in a censored forum that bans people who don't like segwit.  Strange thing that you would find that opinion in such a place...

The facts tell a different story.  In a manually run poll(by humans, not online) Segwit only had 71% of people supporting it.  In a poll of ~70 well known figures in the community, only 75% supported it, the highest of any metric anywhere.  The highest percentage of miners that ever signaled for segwit before segwit2x was 35%.  None of that reached the 80% or 95% desired for activation.

What about your claims of node and business support?  Running a certain type of software is not the same as support, especially when that software is the only one available with a stable history and the others get DDOS'd frequently.  Unless you think that "America supports Trump because they elected him" you need to drop that argument.  Businesses too don't get any better - hence why coin.dance differentiated between being segwit "ready" and "supporting segwit."  Among unweighted companies it was only 49% that _supported_ it before segwit2x, and with weighting that dropped to currently 27%.  Why do they have weighting you ask?  Great question, why does luke-jr alone have 2 "companies" representing 3% of the UASF "vote" both of which are basically defunct projects that no one ever used?  Why would "Yogh.io" be listed, a project that almost no one has ever heard of that hasn't had a code update in over a year?

Even more, before reddit started including non-logged in users in its counts, /r/btc had nearly 30% of the users online that /r/bitcoin had.

> What SegWit2x tries to do is bundle something everybody wants (or can accept) with something only a presumably tiny subsection of the ecosystem wants. 

There's substantial evidence that segwit-only never had consensus, and substantial evidence that segwit2x has that missing consensus.  98% of the miners are now signaling for segwit2x, and segwit2x has nearly double the weighted company support that segwit-only had.  Your claims are based upon nothing but echo-chamber logic that bans competing logic.

> If you dive into the question of why they want it the way they want it, it's usually answered with nothing more than "because".

Have you literally never bothered talking to someone who doesn't agree with you?  Because that would explain a lot.

> If you want to hard fork, then simply create a client that hard forks and see how much support you get. Don't use SegWit as a carrot on a stick.

The irony of this is that using the same logic, segwit would never activate.  Segwit never had consensus, and was not going to get consensus without compromising with the people that they shoved out of the way when they declared they had consensus.

> It's not contentious by any meaningful measure. a) https://en.bitcoin.it/wiki/Segwit_support 

You do realize that recently a core dev pointed to that exact same link claiming the exact same thing, stating that "developers since 2011... support segwit."  And despite that and your claims here, that page REMOVES DEVS WHO HAVEN'T COMMITTED SINCE SEGWIT GOT CONSENSUS BUT ADDS NON-CORE DEVS TO BOLSTER THEIR CLAIMS.  It also has literally 5 shades of green for "support."  How do you not see what is going on here?

> b) in the past 24 hours, 80% of the hashrate signalled for it via BIP91.

Signaled for segwit2x you mean, the compromise that segwit needed to get activated.  I can understand your confusion since core attempted to co-opt the meaning of bit-4 signaling (see Peter Todd's recent emails and "issues") and /r/bitcoin then tried to convince themselves that Bit91 predated the compromise (lol, see the first line of the bip91 email).

Sorry, you've been mislead tremendously.  Maybe you should try reading in places that don't ban competing opinions, and, you know, actually discussing things with the people you are completely dismissing.

> This will be my last reply here, so don't bother asking me further questions.

Or I guess you could just go back to the place where everyone who disagrees you has already been banned.  Definitely will help you mentally resolve the facts I listed above!

**Comment by betawaffle** [comment]
>The irony of this is that using the same logic, segwit would never activate. Segwit never had consensus, and was not going to get consensus without compromising with the people that they shoved out of the way when they declared they had consensus.

If segwit is so disliked, why compromise with s2x?

**Comment by JaredR26** [comment]
> If segwit is so disliked, why compromise with s2x?

Because I don't want two Bitcoins with 42 million coins between them and two communities that hate eachother even more than they already do today.

Many people on the big block community do not want a compromise, and are very angry over it.

The hardfork in November has enough community support to be viable (~40% of miners at least initially, ~25% of the community at least, and as much as 50% of companies [weighted]).  Does it have enough to be a majority in one or more of those ways?  I can't say, not enough data.  Will the legacy chain be viable?  If the signatories stick together, probably not without a hardfork, but if some miners defect, it probably will be viable.

And there's the problem.  Two viable chains competing.  Both with legitimate claims to being Bitcoin.  That's what I'm trying to avoid, and supposedly that's been what *everyone* has been trying to avoid for 2 years.  Unfortunately many of the people who were trying to avoid "forks without consensus" (i.e. core in December 2015, and /r/bitcoin moderation) are now adamantly opposed to the fork that **has** consensus.  They're now going to split Bitcoin just like they said they were trying to stop.  It hasn't been about preventing a split this whole time after all.

**Comment by betawaffle** [comment]
>... are now adamantly opposed to the fork that has consensus.

Consensus among miners is no more consensus than the "lack of consensus" you claim segwit had.

**Comment by JaredR26** [comment]
@Pheromon 

> fee discount for witness is not a consensus rule

FYI, this is not true.  I'm late to this thread so I don't know what has been covered and not.  The fee discount is a consensus rule because it is used(and essential) for the calculation of the blocksize/blockweight within the maximum.  A miner ignoring it would be at risk of producing invalid blocks if their calculation could produce a larger blockweight, and would be leaving fees on the table if they only produced lower weight blocks.

@mjamin 

> In other words you want to block SegWit deployment in order to use it as a bargaining tool to get what you want,

Segwit was used originally to block and derail all blocksize increase proposals.  It has been very effective at that.  It has been effective at achieving consensus in forums that ban people for not agreeing.  Less effective at achieving consensus in a true consensus system like Bitcoin.

@btcdiscounted 

> The reality is the discount is supposed to translate to the transaction price the cost of the utxo vs the cost of the witness data. That is its economic role in the model. 

There's more to it than that.  As I stated in my email linked early on here, the segwit discount actually targets the wrong thing - Some simple math in the example cited on the witness discount FAQ here will show that: https://segwit.org/what-is-behind-the-segwit-discount-8515a8d3bca9

In the example given, segwit does not set things up so that the wallets favor closing UTXO's instead of creating them.  All it does is make them less bad.  The entire purpose of the segwit discount was to give a blocksize increase via soft fork, and some people may have wanted it to favor signature-heavy transactions.

I don't personally see that as a huge problem or bad thing, it just isn't ideal.  Ideal is a witness _adjustment_ that explicitly favors UTXO closing over UTXO opening.  That requires a hardfork to fix, and would have required a hardfork to get it right in the first place (Jeff and many others, I believe, were pushing for segwit to be a hardfork around the time "consensus" was reached on the email list)

@betawaffle 

> > Anyone else has had no saying and the only option if they disagreed is to leave Bitcoin. That is the reality, saying anything else is lying.

> If you believe that, then Bitcoin has already failed at its goals. I don't, but if it's true, Bitcoin was a failure.

Err, dude, you need to re-read what happened between July 2015 and January 31st, 2016.  That is exactly what happened.

> A full node is the only way you can guarantee you're actually receiving Bitcoin that conforms to rules you agree with.

Full nodes are a useless misnomer.  There's several different ways nodes can be run - wallet nodes for holding balances, sybil nodes for making a political statement, transacting nodes for frequent use of Bitcoin, and mining nodes generating blocks.  Of those, sybil nodes serve the least use for the network.  Most UASF nodes appear to be sybil and wallet nodes.

> If you don't run a full node, you're doing little more than trusting the sender. 

This isn't true.  SPV wallets provide substantial security and have numerous advantages over nearly all legacy payment systems.

> Without a full node, Bitcoin isn't trustless.

Running a full node provides a level of security and trustlessness that 99% of the world's population absolutely does not need.  This logic is literally what is preventing Bitcoin from scaling - Using your logic, Bitcoin will never be able to serve even 0.01% of the world's population or worldwide transaction volume.  Congrats on making a coin for the wealthiest 0.01% of the world, but at least its trustless enough for you!

> > What should have happened is that an algorithm based on crypto and economic incentives should have governed the trade off between the utxo cost and the witness cost.

> That is exactly what they did. What makes you think otherwise?

As I stated above, both of you are wrong here.  The witness discount does not target UTXO's, the witness discount is about scaling without a hardfork.  The reason the hardfork was so strongly opposed was about precedent, not because hardforks can't be good.

> The point of the change is to correct the incentives to make running nodes less costly. 

Segwit literally does not do this.  You've been mislead, go do the math if you don't believe me.  It makes running nodes more costly by increasing scale, and it increases bandwidth consumption on the whole network(the biggest cost!) by making signature-byte-heavy transactions less expensive than their bytes would indicate.

> Bigger blocks force users out of the ability to run full nodes

I have literally never gotten a coherent response to this, so I'll ask it here in the hopes that you can answer this...

Why should running a node cost a cup of coffee when transacting on Bitcoin costs rent money?

Both are necessary costs to use bitcoin, but node costs can be spread out across more people with SPV nodes, providing the right level of trustlessness and security for each person and usecase and allowing more than the richest 0.01% of the world to transact.

> run full nodes, which is the only thing protecting them from outside control.

This is not true.  There are many many layers of Bitcoin security.  Full nodes protect against a very limited set of miner collusion attack vectors, but nearly all of those attack vectors are already mitigated by Bitcoin's design and Bitcoin's ecosystem in other much more important ways.

Running a node that doesn't transact or mine just sybiling.  It doesn't increase the security of the network, and it only increases your own security in your head when you don't consider the other layers of protection that Bitcoin has against the things you fear.

**Comment by JaredR26** [comment]
> It's a sliding scale, the larger the blocks, the more people won't be able to afford running a node. 

This is false, studies have shown that as transaction volume increased, the number of nodes increased.  Moreover, the number of transacting nodes can only increase with increased transaction volume.  Sybil nodes are nearly useless.

> Consensus among miners is no more consensus than the "lack of consensus" you claim segwit had.

I provided like 9 different ways where segwit lacked consensus.  You responded to one, the most important one, where segwit2x DOES have consensus.  Is that the best you can do?  Come on, respond to facts with facts.  Despite the several things I cited showing that segwit lacked consensus to activate by any metric picked, you haven't provided anything to back your "lack of consensus" claim regarding segwit2x.

Not to mention that my point wasn't even about miners having the final say on consensus.  My point was that opposing segwit2x is going to give us two viable competing forks, supporting it is the only way to keep Bitcoin whole.

Moreover, miners have a lot more power than you or most anyone in the /r/bitcoin censored forum are willing to admit.  Without sufficient miners, the legacy chain will halt completely.  With 8% of the network hashrate, the first difficulty adjustment will be 6 months out if and only if none of that 8% defect for the entire 6 months.

Note, I'm not stating that the reverse isn't true either - Miners do not have absolute power, no matter what /r/btc people say.  Miners splitting from the market are punished severely and have to take massive risks to attempt to force behavioral changes without near unanimity.  Bitcoin only exists with some nontrivial portion of each group coming together - users, markets/merchants, miners, developers, and even nodes.  Of those, miners (by design) have the strongest control when united but are nearly powerless individually.  Developers on the other hand have substantial control individually (total number is low and code isn't generally reviewed by a substantial number of people on each side of a disagreement), but don't gain any more control by uniting (more developers can always step in).  Nodes play a substantial role in protecting Bitcoin from DDOS attacks and sybil control, and providing a level of routing anonymity, but they have very little say in consensus rules because users, markets/merchants, and miners are all also nodes.

There's a lot of confusion in the community about how the different groups balance eachother and come together.  Intentionally, for political reasons, as each group tries to push their agenda.  I don't fault them for doing that, it is natural, but people need to be honest about the way the Bitcoin ecosystem comes together.

**Comment by betawaffle** [comment]
>Running a node that doesn't transact or mine.

Who is suggesting that? I transact with my node. A node that isn't doing those isn't useful, and only costs resources.

>Why should running a node cost a cup of coffee when transacting on Bitcoin costs rent money?

I don't quite understand this question.

>... with SPV nodes, providing the right level of trustlessness and security for each person and usecase ...

The "right level" is subjective and clearly we disagree on where to draw that line. From my perspective, you want me to give up one of the "layers" of protection by making it impractical for me to run a fully validating node. Just because you don't think more than SPV is necessary doesn't mean all of us should be forced to go that route.

**Comment by betawaffle** [comment]
>This is false, studies have shown that as transaction volume increased, the number of nodes increased. 

The number of nodes doesn't matter (you even agree about that). But *I personally* couldn't run a node if it used all my bandwidth or storage space.

**Comment by betawaffle** [comment]
>... supporting it is the only way to keep Bitcoin whole.

I'm fine with a split Bitcoin if the alternative is a Bitcoin without what I value in Bitcoin.

**Comment by betawaffle** [comment]
The problem is, segwit didn't have "consensus" among miners (obviously, or it would be active), and larger blocks don't have "consensus" among users. You can't disprove that any more than I can prove segwit has user consensus. We are both choosing to interpret either different data, or the same data differently.

None of the three options is going to get all the users.

**Comment by JaredR26** [comment]
> Who is suggesting that? I transact with my node. A node that isn't doing those isn't useful, and only costs resources.

There's whole threads of people on /r/bitcoin setting up raspberry pi's with no intention of ever transacting with them.  rPI bitcoin nodes are literally only usable by the proportion of the population that can use the linux command line, which is a teeny teeny minority of the world.  And they aren't the only ones setting up these sybil-only nodes.  Since people have been convinced that node version/signaling counts for something, there's a motivation for sybiling for political reasons.  BU and classic were guilty of it too (though would have actually needed the DDOS protections that provides if they were to have actually activated any forks).

> I don't quite understand this question.

Ok, let me rephrase it (I would really, really like a response to this, seriously... I have yet to get a coherent counter to its logic):  Looking at what you said a moment later

> But I personally couldn't run a node if it used all my bandwidth or storage space.

Great, and you aren't alone in saying/thinking that.  The problem between what you're saying here and what I said about coffee/rent costs is the scale of the problem.  The world is huge [citation needed] and has many many people [citation needed].  426 Billion transactions worldwide in 2015.  If you actually want Bitcoin to serve more than a few people, it needs to fit those transactions somewhere.  I'll handle lightning in a moment*  because I know that's where many people go mentally right now, so bear with me or scroll down.  426 Billion transactions globally, which itself is increasing by 10% per year, cannot be fit anywhere except a datacenter** as it just consumes too much bandwidth.  But if you can't fit the transactions, many people and use-cases must be cut off from blockchain space.  As I'm sure you know, that cutoff is fees.

So if you keep blocksizes small to have low node operational costs, either fees must rise to cut off use-cases, or Bitcoin simply can't grow.  We're **tiny** right now, 0.02% of the world's transaction volume and probably used by an even lower percentage of the population than that.  If it doesn't grow, Bitcoin isn't really providing freedom and trustlessness for people who want it, is it?  So it must grow.  But if it grows, fees must rise because we've already established blocksizes won't grow.

There's absolutely no limit to how high fees will rise.  If you take the biggest 0.02% of the world's transactions, you're looking at multi-million dollar transactions exclusively.  $1000 fee for a multi-million dollar transaction?  No problemo.  Without increasing blocksizes, **a lot**, fees absolutely can rise to that level and beyond.  So now your node costs the equivalent of $5 a month to run, but each transaction costs upwards of $1000?  This makes no sense.  If you just paid $50 a month to run, transaction fees could be $10 or less instead, and you'd spend less on an entire year of node operation+transactions than a **single** transaction on the limited blockchain.

\* Ok, but lighting solves it, right?  Well, Lightning will help, but lightning absolutely does not solve it.  Lightning comes with a very different set of risks, requirements and tradeoffs that is absolutely not appropriate for every user or every use-case.  Lightning requires senders and receivers to be online at all times.  Lightning may have unforseen sybil consequences as it must form chains.  Lightning funds require watchers to prevent defections which requires fees.  Lightning is vulnerable to default-en-masse attacks where an attacker consumes nearly all available blockchain space and profits when they shouldn't have been able to.  Lightning requires that users pre-fill funding for other people in order for those people to be pay-able on lightning.  Lightning can't be used for large transactions reliably which breaks payment loops back to consumer's accounts, causing money to flow in one direction and require frequent channel refilling.  Lightning will have fees that reflect a certain percentage of Bitcoin tx fees, and predicting lightning's fees in advance is nearly impossible given the complexity of tradeoffs and risks it has.  Lowering Bitcoin transaction fees will lower lightning transaction fees.

Despite its flaws, Lightning is a fantastic solution for time-sensitive payments that happen more than a few times, and it is great for high-frequency payment paths.  I do believe Lightning may be able to offload as much as 50% of the transaction volume in addition to enabling new use cases.  That's great and all, but we don't need 50% to serve the world, we need 5000%.

Lastly, even if I am wrong and lightning offloads much more than that, it still doesn't change my argument.  Both bitcoin users and lightning benefit from lower on-chain bitcoin fees and lower blockspace contention.  So even in that case, it **still** doesn't make sense for a node to cost $5 a month when transaction fees approach $1000.  I'm sure $1000 seems ridiculous to you for transaction fees, but people laughed when I was saying months ago that transaction fees couldn't stay below $1 per transaction long term (they were $0.20 at that time).  Despite them laughing it off, transaction fees went from $0.05 per tx avg to $5 per tx avg within a 12 month period.  And that's not "spam" - transactions per second has followed a fairly predictable +65-80% per year growth for nearly 5 years running, March was just when the ceiling really began to hit people.  Many businesses have changed their Bitcoin donation buttons to Ethereum donation buttons because of fees, behaviors are already changing as use-cases get pushed out.

** But what about datacenter nodes?  Datacenters aren't evil.  They provide a service, something anyone can pay for.  There are hundreds of cloud providers to choose from in hundreds of companies, and thousands upon thousands of datacenters to choose from.  They don't care what you do, and that doesn't mean the government controls your node.  Just like people don't need to run a full node on their phone to transact with Bitcoin on their phone, people don't need a full node in their house to be running a full node, and **most** users don't need the security and trustlessness provided by a full node.  They're fine with SPV nodes.

So now that I've written all of that out... **Why should running a full node cost a cup of coffee when transacting costs rent money?**  Both costs are intricately tied to "using" Bitcoin.  Both costs can be and must be balanced.

>  From my perspective, you want me to give up one of the "layers" of protection by making it impractical for me to run a fully validating node.

No, I absolutely do not want that for you.  I want other people to be able to come into the network in the future without paying what will be, at that time, exorbitant transaction fees just to get started.  How can I say this?  I believe node operational costs can be balanced with transaction fees by doing only one thing differently: Calculating node operational costs in BTC, not dollars.  Ergo, as Bitcoin prices rise, nodes can scale to handle more volume.  As of pre-spike January that was 0.005 BTC per month.  You've been in Bitcoin since at least 2015 it seems, so you can probably afford 0.005 BTC per month for the rest of your life.  I know I can and would gladly do so to help Bitcoin scale and not lose to a coin that better balances node costs and transaction fees.  Will you?

> Just because you don't think more than SPV is necessary doesn't mean all of us should be forced to go that route.

The people who don't want to go that route can simply pay more, and it isn't even a lot more they have to pay!  A mere $800 a month will let us scale well to VISA levels and reach nearly 6% of the world's population and transaction volume, and let Bitcoin scale to a price of over $100k per coin.  Since nearly everyone who cares about this kind of thing are already Bitcoiners that will profit as the price rises, they can easily afford to do this for as long as desired.  There's no need for node operational costs to ever rise **when priced in BTC per month.**  In fact, node operational costs can drop in BTC per month while scaling Bitcoin.  The alternative that you're proposing, as I covered above, is that literally no one new can or will begin using Bitcoin, and it will either be supplanted by a coin that serves users, or it will wind up only being usable by the wealthiest 0.02% of the world.

Are you saying you wouldn't justify paying $800 a month for a full node if each Bitcoin was worth $100,000 and nearly every person you know could use / was using Bitcoin?

> I'm fine with a split Bitcoin if the alternative is a Bitcoin without what I value in Bitcoin.

Many different people value many different things in Bitcoin.  Why is what you value in Bitcoin more important than what other people value in it?

**Comment by Pheromon** [comment]
Thanks @JaredR26 for your posts, really great points, well expressed.

About this:

> > fee discount for witness is not a consensus rule
> FYI, this is not true. I'm late to this thread so I don't know what has been covered and not. The fee discount is a consensus rule because it is used(and essential) for the calculation of the blocksize/blockweight within the maximum

It's my understanding (if not please tell me so I stand corrected) that only the aggregate weight of txs is considered, so it's not so important how much I weight the single transactions but the fact that the aggregate size of witness data must be less than 3MB.

> This is false, studies have shown that as transaction volume increased, the number of nodes increased

More than that, we have ethereum as an example where the node requires *way* more resources than bitcoin, yet there are many times the number of eth nodes than bitcoin. See also http://www.cnbc.com/2017/07/19/ethereum-takes-over-digital-currency.html

> most users don't need the security and trustlessness provided by a full node. They're fine with SPV nodes

For most users even SPV is more than overkill. They can simply use one of the many wallet providers that offers a service to transact with the blockchain while keeping the keys on the client.

**Comment by betawaffle** [comment]
>There's whole threads of people on /r/bitcoin setting up raspberry pi's with no intention of ever transacting with them.

Well those people are dumbasses ;)

>If you actually want Bitcoin to serve more than a few people, it needs to fit those transactions somewhere.

I'm of the impression that cryptography advancements and other things will make Bitcoin more able to handle this in the future, without sacrificing anything. I'm sure you can agree we don't need to handle that much activity yet. It then becomes a line drawing of how much activity do we really need to handle in the short term.

>So it must grow.

I agree that it must grow, but maybe not at the rate you are suggesting. I have not yet made up my mind whether it's possible to control the rate of growth or not, though.

>There's absolutely no limit to how high fees will rise.

Well, there certainly are limits, but they may be very high. As fees increase, more of the money becomes too costly to move, which could drive up the price of Bitcoin, which could reduce fees. Just a theory, though.

I'm going to reply to the rest in smaller comments, because there's so much.

**Comment by christophebiocca** [comment]
> As fees increase, more of the money becomes too costly to move, which could drive up the price of Bitcoin, which could reduce fees.

I'm assuming you meant drive *down* the price of Bitcoin? Otherwise that just makes things worse.

**Comment by JaredR26** [comment]
> and larger blocks don't have "consensus" among users.

Everything I can find outside of /r/bitcoin upvotes indicates that this is false.

> You can't disprove that any more than I can prove segwit has user consensus

Prove?  No, but I can provide a lot of data that you have to explain somehow.

Manually run by humans to prevent sybiling, 64% support: https://cointelegraph.com/news/the-bigger-the-better-64-of-bitcoin-users-vote-for-block-size-increase

Voted by address signatures, not sybilable: https://vote.bitcoin.com/arguments/block-size-limit-should-be-increased-to-8-mb-as-soon-as-possible

and: https://vote.bitcoin.com/arguments/bip101-bip102-and-bip103-increase-scale-without-increasing-scalablity-potentially-a-cost-to-security

Proof-of-twitter poll... Also note how laughably small these result sets are... If nothing else, it should terrify you that Bitcoin is so small that a contentious sybilable poll only gets less than 3000 votes: https://twitter.com/ViaBTC/status/860933065022898176?s=09

Sybilable poll from 2015, which **also** got a teeny tiny number of votes:
http://blog.genesis-mining.com/survey-results-increasing-the-blocksize

Big blocker support nearly 15% higher than small blocker support, not sybilable (known names): https://medium.com/@21/using-21-to-survey-blockchain-personalities-on-the-bitcoin-hard-fork-1953c9bcb8ed

Even Luke-Jr's sybil-resistant poll, heavily slanted by the questions Luke picked to support his case and essentially ignored by people outside /r/bitcoin shows that more people support 4mb blocks by 2019, something segwit cannot give us (theoretical maximums that can't be hit under realistic circumstances aside), but **still** reflects an increase in node operational costs faster than technological improvements.

I haven't found any polls anywhere, sybilable or not, suggesting that small blocker support is stronger than big blocker support, the closest being Luke's.

> We are both choosing to interpret either different data, or the same data differently.

I mean, I provided a bunch of data just now... can you do the same?  Otherwise I don't think that's a fair claim to make.  I would agree with you that there's a lot of not-very-helpful data out there, at least.

> None of the three options is going to get all the users.

This is totally true, I agree with you!  Fortunately that's a ridiculous goal, and the only people who are setting that goal (hint on who that is...) have an ulterior motive - their ACTUAL goal is maintaining the existing consensus rules, they could care less what other people want or how many of them there are.

Fortunately for us, Bitcoin has strong forcing functions to keep the extreme minorities in line - Their forks won't be viable and will fail.  The real threats worth discussing are viable forks, and, I think, broad support for proposals (in communities where competing viewpoints are allowed or hopefully encouraged, and no I wouldn't consider /r/btc to be a place "encouraging" competing viewpoints).

**Comment by Pheromon** [comment]
> There's absolutely no limit to how high fees will rise.

Not really.
Since the fees are high for an artificial limitation and not for a technical one, when fees are too high, what happens is that (new and existing) users and (new and existing) use cases migrate to other coins that has no production quota. This is already happening massively in Bitcoin, and for every user or service that migrate from bitcoin, there is value transfered elsewhere.

**Comment by betawaffle** [comment]
I can't dispute what you've said about LN so far, so I'll move on.

>Many businesses have changed their Bitcoin donation buttons to Ethereum donation buttons because of fees.

They are welcome to do so, I don't see why that's such a big deal. If they only care about fees and users don't care then they don't need Bitcoin.

>But what about datacenter nodes?

There is nothing wrong with datacenter nodes.

>You've been in Bitcoin since at least 2015 it seems, so you can probably afford 0.005 BTC per month for the rest of your life.

2011, but that is not at all why I'm able to afford running a node.

>The people who don't want to go that route can simply pay more, and it isn't even a lot more they have to pay! A mere $800 a month will let us scale well to VISA levels

$800/month is not reasonable for an individual. No bank costs that much, why should I pay so much to get the same protections as anyone else?

>... when priced in BTC per month.

I don't earn money in BTC, so how could I afford that?

More to come...

**Comment by betawaffle** [comment]
>Are you saying you wouldn't justify paying $800 a month for a full node if each Bitcoin was worth $100,000 and nearly every person you know could use / was using Bitcoin?

Correct.

**Comment by betawaffle** [comment]
>Many different people value many different things in Bitcoin. Why is what you value in Bitcoin more important than what other people value in it?

It isn't, but I'm still going to fight for it.

**Comment by JaredR26** [comment]
> It's my understanding (if not please tell me so I stand corrected) that only the aggregate weight of txs is considered, so it's not so important how much I weight the single transactions but the fact that the aggregate size of witness data must be less than 3MB.

No, I had this wrong as well, very publicly at one point (oops!).  Your statement implies that there are two limits for segwit, and there are not.  There's one line - maxblockweight - and the formula that calculates maxblockweight just happens to be set at the point where the original max base blocksize couldn't be over 1mb.

This, of course, makes calculating how much actual bandwidth Bitcoin consumes much more difficult for humans.  Minor downside, but not exactly a desirable trait.

**Comment by JaredR26** [comment]
> I'm of the impression that cryptography advancements and other things will make Bitcoin more able to handle this in the future, without sacrificing anything

Your impression is wrong.  Even if we got Bitcoin transactions down to a ridiculously small(and probably extremely insecure) 50 bytes of data each we could not serve the world's transactions without increasing node operational costs.

I'm not saying we can't do better or that we shouldn't, my point is that that logic simply doesn't work for the scale of the problem.

> I'm sure you can agree we don't need to handle that much activity yet.

No, we need to handle +65% - +80% per year.  That's what the transaction growth naturally drives towards historically.  More accurately, I think we need to increase blocksizes whenever the average fee per transaction rises above $1-2 per tx.

> I have not yet made up my mind whether it's possible to control the rate of growth or not, though.

Markets don't like attempts at control.  Ethereum would not be at 50% of Bitcoin's market cap if not for the scaling problem.  No Bitcoiner can prevent that growth, and many support it as they sell Bitcoins to buy Ether.

> As fees increase, more of the money becomes too costly to move, which could drive up the price of Bitcoin, which could reduce fees. Just a theory, though.

Wait, did you just say as fees increase, fees will decrease?  :P

Fees drop when either usage drops or when blocksizes are increased.  Reducing usage of Bitcoin is going to be punished very harshly by the markets that we cannot hope to control.

**Comment by Pheromon** [comment]
> There's one line - maxblockweight - and the formula that calculates maxblockweight just happens to be set at the point where the original max base blocksize couldn't be over 1mb.

Yes, but my point is that a miner could decide to change the formula for the fees needed to include a segwit transaction in the block, and that would not violate consensus rules.

For example, a miner could decide to accept segwit transactions only if they pay twice the fees than normal transactions: they would lose money (fees), but that would not violate consensus.

As an hyperbole, a miner could decide to include in its blocks only normal (non-segwit) txs by considering the weight "infinite".

**Comment by betawaffle** [comment]
>I mean, I provided a bunch of data just now... can you do the same? Otherwise I don't think that's a fair claim to make. I would agree with you that there's a lot of not-very-helpful data out there, at least.

I might be able to if I cared more, but I think we agree enough for me to let this point go.

>Even if we got Bitcoin transactions down to a ridiculously small(and probably extremely insecure) 50 bytes of data each we could not serve the world's transactions without increasing node operational costs.

You're assuming we'll need to store the transactions at all ;)

>As fees increase, more of the money becomes too costly to move, which could drive up the price of Bitcoin, which could reduce fees. Just a theory, though.

- Fees increase (more fee per tx) when blocks become more full.
- When fees increase, transaction outputs which are too low to afford the fee are not worth spending.

**Comment by JaredR26** [comment]
> If they only care about fees and users don't care then they don't need Bitcoin.

This is literally the logic that markets punish severely.  Your logic appears to be "it doesn't matter if other people want to use other things more than Bitcoin."  But it does, crypto-currencies are networks, and network effects matter a whole hell of a lot.  Myspace and friendster used the same logic at one point.

> $800/month is not reasonable for an individual. 

I never said it was.  The vast majority of individuals get all the security they need between SPV clients and light clients like Electrum validating from multiple sources.  Fullnode operation can be run as a service and amoritize node costs across thousands of users, just like pool nodes already do today for miners.

> No bank costs that much, why should I pay so much to get the same protections as anyone else?

Banks can easily cost that much if you want protections that other people don't get, like financial advisors watching your investments for shifts and trends 24/7.

Not every person needs the same level of security or services.  You're imagining that "risk" is an absolute, it is not.  Did you review every line of code and recompile your Bitcoin node from that source code base?  If not, you have taken what 99.9% of Bitcoiners would call a reasonable risk.  The viable attacks against SPV clients are extremely expensive to perform and have an extremely low chance of impacting the average user.  Far far less if they also use multiple verification from light client services.  These aren't hypothetical, this type of thing could be done TODAY and let Bitcoin scale to reach hundreds of millions of people.  Let people choose the risk levels that are appropriate for their usecases without preventing other people from using Bitcoin _at all_.

> I don't earn money in BTC, so how could I afford that?

Assuming you are 15 years old today or older and live to be 90, if you have 4.5 BTC today you could afford to run a full node like this for the rest of your life.  I believe you should have the choice **not** to do that if you do not want to, but your choice should not simultaneously prevent huge numbers of other people from **_ever using bitcoin._**

**Comment by betawaffle** [comment]
>The vast majority of individuals get all the security they need between SPV clients and light clients like Electrum validating from multiple sources. Full node operation can be run as a service and amortize node costs across thousands of users, just like pool nodes already do today for miners.

What about privacy?

>Assuming you are 15 years old today ...

When you assume, you make an ass of u and me.

**Comment by betawaffle** [comment]
>Let people choose the risk levels that are appropriate for their use cases without preventing other people from using Bitcoin at all.

Bitcoin can't compete with alternatives that have lesser privacy and trustlessness considerations. To lower the accessibility of those properties just to bring in more users immediately isn't worth it, IMO.

**Comment by JaredR26** [comment]
> > Are you saying you wouldn't justify paying $800 a month for a full node if each Bitcoin was worth $100,000 and nearly every person you know could use / was using Bitcoin?

> Correct.

Then either you don't hold enough Bitcoins for Bitcoin's economic incentives to actually motivate you the way they were designed to (why are you here if this is the case?) or else you're very stingy.  I'd donate 100 BTC if it let Bitcoin scale to reach hundreds of millions of people, and I joined around the same time as you.

> It isn't, but I'm still going to fight for it.

And to you, that is worth splitting Bitcoin into two pieces and doubling the total number of coins in circulation, and permanently harming Bitcoin's image as a stable, reliable, secure currency?  That's a pretty harsh tradeoff if you're successful, not to mention the millions of people you exclude from ever using Bitcoin.

> > Even if we got Bitcoin transactions down to a ridiculously small(and probably extremely insecure) 50 bytes of data each we could not serve the world's transactions without increasing node operational costs.

> You're assuming we'll need to store the transactions at all ;)

No, I'm literally not.  426 billion transactions per year = 35.5 billion transactions a month = 1.775 trillion bytes of data a month if each transaction is received only once and never relayed.  That's 1775 GB a month of bandwidth raw.  Factor in transaction relay requirements(can't just receive!) and network relaying inefficiencies and its more like ~14 terabytes of bandwidth a month, $250 a month of bandwidth using about a 10th percentile of the best bandwidth prices available today(0.019 per GB).  The smallest transaction is a bit over double that.

And if you don't store transactions btw, you aren't contributing to historical network sync, which is needed for absolute trustlessness as you are pushing for here.

As an aside, I agree with that statement in one way.  We desperately need a fast hashed-utxo sync or checkpointing system like what has been implemented in nearly every other cryptocurrency of any significant size.  That hasn't been prioritized much like compact blocks wasn't prioritized until BU did it. Things that reduce raw bandwidth consumption harm the arguments being put forth by small-block proponents.

> When fees increase, transaction outputs which are too low to afford the fee are not worth spending.

So you are saying you support stealing money from poor bitcoiners by making their UTXO's useless?

What?

**Comment by betawaffle** [comment]
>That's 1775 GB a month of bandwidth raw.

You're making my point for me.

For the most part, we are not disagreeing on any facts, just where to draw the lines.

>And if you don't store transactions btw, you aren't contributing to historical network sync, which is needed for absolute trustlessness as you are pushing for here.

I can only assume you don't know what direction the research is headed.

>We desperately need a fast hashed-utxo sync or checkpointing system.

I agree, but the security implications are still being worked out. You can't just shove things in because it *seems* like it will work correctly.

>So you are saying you support stealing money from poor bitcoiners by making their UTXO's useless?

Of course not. They wouldn't be able to spend them. Fees don't increase forever.

**Comment by betawaffle** [comment]
>I'd donate 100 BTC ...

If I had 100 BTC I'd be a lot more focused on getting the price to rise too.

**Comment by JaredR26** [comment]
> What about privacy?

Privacy can be achieved by balancing balance requests across multiple full node datasources and adding in spurious balance requests for addresses not yours.  Very easy solution that doesn't require hamstringing the entire ecosystem's growth.

People who want even more privacy than that absolutely should be able to do so.  I fully support people who want to run their own tor nodes and bitcoin nodes being able to do so, right up until they decide that their tiny need for privacy is more important than other people being able to **use the currency in the first place**.

> When you assume, you make an ass of u and me.

Very funny, so you're saying my assumption of the worst case is actually not the worst case, and you're younger than that?  I am using the most extreme-but-still-reasonable examples in your favor with both the bandwidth/transaction calculations and also the costing calculations.

> Bitcoin can't compete with alternatives that have lesser privacy and trustlessness considerations. 

Neither you nor I get to decide that.  The markets decide that.  The markets have already spoken that scaling is more important.

> To lower the accessibility of those properties just to bring in more users immediately isn't worth it, IMO.

Again, YOUR OPINION.  The users that you are refusing to allow to **use the coin in the first place** would disagree, and they will disagree, and they will drive up the price, value, technology, and security of alternatives to Bitcoin instead if you don't let them in.

I strongly support your right to have and express your opinion and have it considered critically.  I also support Bitcoin offering you all of the features, security, and privacy that you want, right up until offering you a little more privacy for less money prevents Bitcoin from offering any features, security, or privacy **at all** to other people.  Sorry, you and I aren't special.  The markets will decide, and the markets are absolutely going to punish all of us if Bitcoin splits.

**Comment by betawaffle** [comment]
>... and you're younger than that?

I'm obviously 5 years old.

>Sorry, you and I aren't special. The markets will decide, and the markets are absolutely going to punish all of us if Bitcoin splits.

And what if the market decides that the 2x hard fork isn't the way to go? What will you think then?

**Comment by JaredR26** [comment]
> For the most part, we are not disagreeing on any facts, just where to draw the lines.

Right, and that's great.  But ultimately we either have to come to an agreement about where to draw those lines (Compromise, i.e., segwit2x) or we have to split permanently and accept all the bad things that follow from that.

> I agree, but the security implications are still being worked out. 

Worked out?  They've been actively in use in other coins for **years**.  Meanwhile, there's almost no posts about the issue on the core dev email list over the last 2 years.  One well-known core dev explicitly stated that it "hasn't been a priority," though I've lost the link.  That's all I'm saying - these things we desperately need aren't a priority so long as they don't help the vision that is being pushed politically.

> If I had 100 BTC I'd be a lot more focused on getting the price to rise too.

Fair enough, but that seems to imply either that your reasons for your position already are, or would become, motivated for selfish reasons, no?

There's nothing wrong with that, that's how Bitcoin was designed.  But we need Bitcoin to scale regardless, otherwise its just you wanting to spend less on privacy at the cost of other people even being able to use Bitcoin at all.  We should be allowing different scaling proposals to compete freely instead of being silenced in major forums.

> And what if the market decides that the 2x hard fork isn't the way to go? What will you think then?

It won't decide that for many months because of the strength and philosophy of the groups pushing for bigger blocks, but if it did eventually favor smaller blocks?  Then I would stick with smaller blocks until the blocksize limit becomes a huge pain again and this time the small blockers wouldn't have an excuse to avoid the increase that they could sell to the moderates, and lightning wouldn't be able to hide its shortcomings behind the lack of segwit's activation.

This issue isn't going to magically go away, it is just going to get worse as other coins without this arbitrary limit gain more and more merchants, exchanges, users, devs, nodes, and miners.  Once a use-case switches to using an alt-coin, they generally won't switch back without a powerful motivation.  Every single one we lose is a big problem long term.  But sure, lets just split Bitcoin into two weaker halves and trash everything so you can pay $5 per month instead of $10 per month in node costs.  👍

**Comment by betawaffle** [comment]
>Fair enough, but that seems to imply either that your reasons for your position already are, or would become, motivated for selfish reasons, no?

My interest in Bitcoin is the technology and future potential, not the price.

>We should be allowing different scaling proposals to compete freely instead of being silenced in major forums.

Reddit is not a place for any reasonable discussion on the topic.

>But sure, let's just split Bitcoin into two weaker halves and trash everything so you can pay $5 per month instead of $10 per month in node costs.

I honestly don't think that will be the case.

**Comment by JaredR26** [comment]
>  My interest in Bitcoin is the technology and future potential, not the price.

Still the same... You want Bitcoin the technology and future potential to only be that potential for 0.02% of the population?

> Reddit is not a place for any reasonable discussion on the topic.

Agreed there.  But then again, nowhere seems to be.

> I honestly don't think that will be the case.

With what part?

**Comment by betawaffle** [comment]
>With what part?

That miners will split away from all the ~100k core nodes, or that those nodes will "upgrade" to btc1.

**Comment by JaredR26** [comment]
> That miners will split away from all the ~100k core nodes, or that those nodes will "upgrade" to btc1.

They said they would and signaled it.  You're right that some could stop.  But the rest have been itching for a split for a long, long time, this is past due for them.

> the ~100k core nodes, 

Also you seem to believe running a core node is a vote of something.  As I covered above, it is not.  Donald Trump is president of the United States, ergo Americans agree he was the best pick for president?  Critical thinking man.

Even more than that, how do you explain 100,000 "core supporting nodes" who will adamantly cling to an unusable chain over their philosophy when an easily sybilable poll has less than 3000 votes and 56% big blockers?  Do you have ANY way there to resolve the obvious discrepancy between what you are claiming and the data available?

There's three different ways this could go.  In only one of those cases do we not end up with either 2x or two coins, and maybe not even then.

1. Miners hold to agreement; Legacy chain is not viable with ~1-10% of hashrate.  We either end up with segwit2x or (if core pushes PoW change) with two coins.

2. Miners defect similar to previous BU/segwit support levels, and the userbase splits.  In this situation most likely the segwit2x coin will begin to change quickly since segwit will be much more unpopular among its userbase.  Definitely two coins here.

3. Users refuse to switch because reasons and hold do it for an extended period of time.  Under this scenario Bitcoin would have two essentially useless forks for some time, one with companies and miners but no users, and one with users but no blocks.  Price on both would tank under $500 easily.  Things might reconverge to one coin or we might have two coins.

Or, a wide majority of miners don't follow the hardfork.  In that case we would have one Bitcoin but might still have a much smaller ETC like scenario.  Two coins still possible.

But sure, I'm sure all of those 100,000 users(* we don't really have 100k nodes btw) are frothing-at-the-mouth core supporters who hate miners so much that they are willing to stick with an unusable chain for months just to teach the miners a lesson. 👍

**Comment by CosmicHemorroid** [comment]
@JaredR26 Seriously, enough with your propaganda and turning GitHub into Reddit.

**Comment by JaredR26** [comment]
@CosmicHemorroid 

> Feel free to continue the discussion here - closing the issue as it's not something we will likely address.

From jgarzik.  I assumed if other people took it down that path for hours on end, it wasn't a problem for me to respond in kind.

If it was propaganda, you could refute my facts or my logic.

**Comment by betawaffle** [comment]
>If it was propaganda, you could refute my facts or my logic.

@CosmicHemorroid

Don't bother, I tried. 🤕

**Comment by mpatc** [comment]
@cosmichemorroid

Could you try to keep it professional?

Thanks!

On Jul 20, 2017 5:34 PM, "Andrew Hodges" <notifications@github.com> wrote:

If it was propaganda, you could refute my facts or my logic.

@CosmicHemorroid <https://github.com/cosmichemorroid>

Don't bother, I tried. 🤕

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/btc1/bitcoin/issues/92#issuecomment-316836721>, or mute
the thread
